013-generalize-address-spaces.md

```diff
@@ -2,429 +2,372 @@
+like context dependence, pointer size, or embedded access patterns.
+
+**Insert basic example of OMP mapping to GPU memory.
+
+## Proposed solution
+
+Add a new attribute `DW_AT_address_space` to pointer and reference
+types.  This allows the compiler to specify which address space is
+being used to represent the pointer or reference type.
+
+**For discussion** This probably needs more explanation. I wrote about
+it in my reply to Cary. I think this basically requires Baris's
+refined type proposal and this attribute should be applied to all GPU
+pointers. I believe that this is consistent with Tony's original
+idea. I think we need an example in Appendix D showing how this should
+work.
+
+Add a new `DW_OP_mem` operation defined to create a memory location
+description from an address and address space. It can be used to
+specify the location of a variable that is allocated in a specific
+address space.
+
+**For discussion** Why isn't this a flavor of addr. We probably need
+to cover relocation here.
+
+**For Discussion** Is the OpenCL reasoning that Tony cited compelling
+enough that the address space can't be an operand. If so we should
+probably state that in the document.
+
+Implicit conversion back to a value is limited only to the default
+address space to maintain compatibility with DWARF 5\. This approach
+of extending memory location to support address spaces, allows all
+existing DWARF 5 expressions to have the identical semantics.
+
+Having the address space within the location also allows great
+implementation freedom. For example, a compiler could choose to access
+private memory when mapping a source language thread to the lane of a
+wavefront in a SIMT manner. Or a compiler could choose to access the
+same private memory by mapping the same language construct with a
+wavefront being the thread. It also allows the compiler to mix the
+address space it uses to access private memory. For example, the
+compiler can still spill entire vector registers to an address space
+without a strided access pattern, while accessing the same bits for
+SIMT variable access using the strided address space.
+
+Having address spaces within the location also makes it available to
+consumers and therefore developers. This can also be an aid to
+programmers who are seeking to optimize their code.
+
+Since memory locations are an abstraction of storage. The same set of
+operations can operate on memory locations independent of their
+underlying storage. Therefore, operations like `DW_OP_deref*` can be
+used on any memory locations even those referring to different address
+spaces. Consequently, the `DW_OP_xderef*` operations are unnecessary,
+except as a more compact way to encode a non-default address space
+address followed by dereferencing it. Therefore we propose renaming
+those operators to `DW_OP_aspace_deref*` to be consistent with the
+other operators which refer to address spaces and making the old
+`DW_OP_xderef` names aliases.
```

Comment:
Suggest explicitly mentioning that the necessary changes to support address spaces in CFI will be submitted as a separate issue.